accepted/2022/version-selection.md

```diff
@@ -0,0 +1,449 @@
+
+There are two aspects we need to consider for this experience:
+
+- Repos should be able to require the use of a private install.
+- This feature needs to be opt-in, per the principles stated earlier, and to ensure that the workspace is explicitly trusted by the user (similar to [VS Code Workspace Trust](https://code.visualstudio.com/docs/editor/workspace-trust)).
+
+"Convention location" characteristics:
+
+- Only supported for the `.dotnet` directory (not configurable to something else).
+- The directory must contain an SDK installation, otherwise error. The user will be directed to the [dotnet-install script](https://docs.microsoft.com/en-us/dotnet/core/tools/dotnet-install-script) help page.
+- Can be in current directory or (grand-*)parent (same probing logic as `global.json`).
+- The presence of `.dotnet` stops searches for `global.json`.
+- If `.dotnet` and `global.json` exist in the same directory, `.dotnet` wins; `global.json` version information is ignored, however other information may still be used.
+- `global.json` can require using the "convention location" feature, with a new `useLocal` boolean property (child of `sdk`). The `.dotnet` directory must exist at the same location.  
+- User must opt-in to using a convention location, even if it is required by `global.json`.
+- No relationship to `nuget.config`.
+
+User opt-in characteristics:
+
+- `dotnet` will offer a new command to trust a repo: `dotnet workspace trust [repo-location]`.
+- It is an error if a `.dotnet` directory is found and this command has not been run.
+- Workspace trust data will be stored at `%USERPROFILE%/.dotnet/workspace` using a similar scheme as VS Code does with its `workspaceStorage`.
+
+Overrides:
+
+- User can override the need to trust a repo by setting `DOTNET_USE_LOCAL` to true.
+- If `DOTNET_USE_LOCAL=true` and `.dotnet` directory is not found, it is not an error.
+- User can override using a local `.dotnet` directory (even if `global.json` requires it) by setting `DOTNET_USE_LOCAL=false` or by passing `--ignore-local` as an argument to `dotnet`.
+
+Notes:
+
+- This experience as written doesn't define or validate the behavior in Visual Studio. That needs more consideration. In theory, it should be transparent to Visual Studio.
+- Apphost will not honor this value, nor does it consult `global.json` in any scenario. Same with the `dotnet myapp.dll` pattern.
+
+Considerations for the future:
+
+- We could add a `dotnet install` command to enable private acqusition.
+- We could install .NET (local) Tools into the repo-resident `.dotnet` directory instead of into the user profile location. That approach might avoid the need for a local tools manifest.
+- It would be straightforward to build a version manager (like [DNVM](https://www.codeproject.com/Articles/1005145/DNVM-DNX-and-DNU-Understanding-the-ASP-NET-Runtime)) using symlinks for the `.dotnet` directory. This would be particularly useful for preview and nightly builds.
```

Comment:
symlinks are [generally dangerous](https://docs.microsoft.com/en-us/windows/security/threat-protection/security-policy-settings/create-symbolic-links#security-considerations) They're meant to be created only by trusted admins, not normal users. You can't rely on a user's ability to create them.

Reply:
I have no trouble creating symlinks on Linux. If we build this feature, we may have to consider creating a .NET-specific feature (use this `.dotnet` at this location).

Reply:
On Windows creation of symlinks to protected folders (such as ProgramFiles) requires the Administrator privileges, which is often not available to developers in enterprise settings.

Reply:
The intention here is userspace -> userspace.